lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1482 @@
+    Set-CharmState -Namespace "novahyperv" -Key "local_ip" -Value $lip
+    return
+}
+ 
+
+function Get-CherryPicksObject {
+    $cfgOption = Get-JujuCharmConfig -Scope 'cherry-picks'
+    if (!$cfgOption) {
+        return @{}
+    }
+    $ret = @{
+        'nova' = @();
+        'networking-hyperv' = @();
+        'compute-hyperv' = @();
+        'neutron' = @();
+        'os-win' = @()
+    }
+    $splitCfgOption = $cfgOption.Split(',')
+    $validProjects = @('nova', 'networking-hyperv', 'neutron', 'compute-hyperv', 'os-win')
+    foreach ($item in $splitCfgOption) {
+        $splitItem = $item.Split('|')
+        if ($splitItem.Count -ne 4) {
+            Throw "ERROR: Wrong 'cherry-picks' config option format"
+        }
+        $projectName = $splitItem[0]
+        if ($projectName -notin $validProjects) {
+            Throw "ERROR: Invalid git project name '$projectName'"
+        }
+        $ret[$projectName] += @{
+            'git_url' = $splitItem[1];
+            'branch_name' = $splitItem[2];
+            'commit_id' = $splitItem[3]
+        }
+    }
+    return $ret
+}
+
+
+function Initialize-GitRepository {
+    Param(
+        [string]$BuildFolder,
+        [string]$GitURL,
+        [string]$BranchName,
+        [array]$CherryPicks=@()
+    )
+
+    Write-JujuLog "Cloning $GitURL from $BranchName"
+    Start-ExecuteWithRetry { Start-GitClonePull $BuildFolder $GitURL $BranchName }
+    foreach ($commit in $CherryPicks) {
+        Write-JujuLog ("Cherry-picking commit {0} from {1}, branch {2}" -f
+                       @($commit['commit_id'], $commit['git_url'], $commit['branch_name']))
+        try {
+            Start-ExternalCommand {
+                git -C $BuildFolder fetch $commit['git_url'] $commit['branch_name']
+                git -C $BuildFolder cherry-pick $commit['commit_id']
+            }
+        }
+        catch {
+            # If the same cherry-pick is applied twice it will error.
+            # This catch is made so the hook will not fail on that error.
+            Write-JujuWarning "Git cherry-pick has errored. This is probably because you are trying to cherry-pick the already applied patch."
+        }
+    }
+}
+
+
+function Initialize-GitRepositories {
+    Param(
+        [ValidateSet("hyperv", "ovs")]
+        [string]$NetworkType,
+        [string]$BranchName,
+        [string]$BuildFor
+    )
+
+    Write-JujuLog "Cloning the required Git repositories"
+    
+    $standardProjects = 'openstack/nova', 'openstack/neutron', 'openstack/os-win'
+    $cherryPicks = Get-CherryPicksObject
+    $openstackBuild = Join-Path $BUILD_DIR "openstack"
+    if ($NetworkType -eq 'hyperv') {
+        if ($BuildFor -ne "openstack/networking-hyperv") {
+            Initialize-GitRepository "$openstackBuild\networking-hyperv" $NETWORKING_HYPERV_GIT "master" $cherryPicks['networking-hyperv']
+        }
+    }
+
+    if ($BuildFor -eq "openstack/nova") {
```

Comment:
We discussed this at least 3 times! Instead of all those useless ifs, you do: if ($BuildFor -ne "openstack/nova") {install nova from repo} and same for the others!